openmls/TEST-DESIGN.md

```diff
@@ -0,0 +1,175 @@
+# OpenMLS testing proposition
+
+The purpose of this document is to improve tests in openmls and try to agree on a common standard for writing them.
+
+## Current status
+
+Here are some factual remarks about the current state of tests.
+
+1. erratic: no clear pattern emerges. It looks like tests have been piled up at different times by different persons.
+   There are very few unit tests, some integration tests in `src/` and some other integration tests in `tests/`
+2. too verbose: some domain are of course complex and hard to test, but some simpler ones require thousands of lines to
+   get tested
+3. not idiomatic: one of the recommended way to write tests in Rust is to write unit tests in the same file as the
+   sources in order to have access to private members. Integration tests (in `tests/`) are more high level and
+   especially
+   used for verifying the public API
+4. Not enough factorization: it is not uncommon to have a file redefining its own set of helpers. A lot of those helpers
+   are duplicated all across the codebase
+5. some tests actually test more than one thing. It makes them harder to write/understand. Also harder to debug when
+   they
+   fail (in which iteration am I ?). This results mostly from the test setup being verbose which incentivizes reuse
+
+## Goals
+
+This proposition aims at:
+
+1. Write less verbose tests and quicker
+2. Share a common pattern for writing tests
+3. Have finer grained tests and be quicker to troubleshoot the cause of a failing test
+4. Be idiomatic to welcome easier external contributions
+
+## Rules
+
+To achieve these goals, here are the "rules"/recommendations to set up:
+
+1. Any test expected to fail by altering a valid input should also keep this valid input and verify that, contrary to
+   the
+   tampered one, it succeeds.
+2. All errors should be namely asserted i.e. `assert!(result.is_err())` is prohibited. When matching an error, one
+   should `assert!(matches!(result.unwrap_err(), MlsError::MyError))` because errors are not supposed to impl Eq (it can
+   impose hard to fulfill constraints on the data they wrap)
+3. Prefer `.unwrap()` over `.expect("...")`. The latter adds overhead and most of the time does not add enough debugging
```

Comment:
OpenMLS currently has the opposite. What's the rationale behind this?
Generally, when there's information to be given, it should be given in an `expect`. So I don't agree with the general `prefer` statement.

Reply:
The idea is that most of the time, the reason why those Result panic is straightforward, even without looking at the backtrace. Having `expect` bloats the method a bit and most of the time there's no information to be given.  

There seemed to be a rule of thumb in openmls to always use `expect` in tests. It is also a common practice to prefer `unwrap` so it can only make the code more idiomatic